### DISCUSSION TITLE
Recursion error during numerical evaluation of box diagram

### DISCUSSION THREAD

**Discussion by tonymenzo** [FeynCalc/feyncalc#105] [Q&A]
For background - I am attempting to numerically compute one loop amplitude contributions in the mass insertion approximation for a one-to-three decay process. To get a feel for the algorithm I am computing the decay rate for just one diagram. My idea was to use FeynCalc's `TID` function to first rewrite the amplitude (which I have written out explicitly) in terms of Passarino-Veltman scalar integrals and then use the following substitution rule 
```
toLT := {FeynCalc`PaVe[i__Integer, {p___}, {m___}, O___] -> 
    LoopTools`PaVe[i, ToExpression["{" <> ToString[p] <> "}"], 
     ToExpression["{" <> ToString[m] <> "}"]], 
   FeynCalc`D0[X___] -> LoopTools`D0[X], 
   FeynCalc`C0[X___] -> LoopTools`C0[X] , 
   FeynCalc`A0[X___] -> LoopTools`A0[X], 
   FeynCalc`B0[X___] -> LoopTools`B0[X] };
```
to evaluate these functions numerically using LoopTool's. This seems to work as expected, after substituting the relevant numerical factors I end up with a lengthy summation of spinor-gamma matrix blocks. The next natural step, in my mind, is to obtain the spin-averaged squared amplitude using the `FermionSpinSum` function. Upon doing so I seem to run into an infinite loop with errors
```
$RecursionLimit::reclim2: Recursion depth of 1024 exceeded during evaluation of \[Not]OptionQ[{}].
$RecursionLimit::reclim2: Recursion depth of 1024 exceeded during evaluation of {}==={}.
$RecursionLimit::reclim2: Recursion depth of 1024 exceeded during evaluation of Hold[\[Not]OptionQ[{}]].
General::stop: Further output of $RecursionLimit::reclim2 will be suppressed during this calculation.
$IterationLimit::itlim: Iteration limit of 4096 exceeded.
$IterationLimit::itlim: Iteration limit of 4096 exceeded.
$IterationLimit::itlim: Iteration limit of 4096 exceeded.
General::stop: Further output of $IterationLimit::itlim will be suppressed during this calculation.
```
I have been unable to find a solution to this problem. My thought is that maybe FeynCalc's internal language is getting tangled when I evaluate PaVe's with LoopTools. I have not been able to find the origin of the `OptionQ[...]` even after looking at the stack. 

I have attached a notebook which contains my code. If you have any suggestions/corrections/comments on my implementation they would be much appreciated. Also, if you need any other information let me know.

I am using Mathematica 12.1 with FeynCalc 9.3.1 and LoopTools 2.15 on a Manjaro system. 

[recursiveErrorBoxDiagram.zip](https://github.com/FeynCalc/feyncalc/files/6206119/recursiveErrorBoxDiagram.zip)

**Comment by vsht** [comment]
As a quick idea (I didn't look at the code yet) you may try
```
toLT := {FeynCalc`PaVe[i__Integer, {p___}, {m___}, O___] -> 
    LoopTools`PaVe[i, ToExpression["{" <> ToString[p] <> "}"], 
     ToExpression["{" <> ToString[m] <> "}"]], 
   FeynCalc`D0[X___,OptionsPattern[]] -> LoopTools`D0[X], 
   FeynCalc`C0[X___,OptionsPattern[]] -> LoopTools`C0[X] , 
   FeynCalc`A0[X___,OptionsPattern[]] -> LoopTools`A0[X], 
   FeynCalc`B0[X___,OptionsPattern[]] -> LoopTools`B0[X] };
```

BTW, I had an idea of a proper LoopTools interface via FeynHelpers,
but it got stalled due to a lack of public interest and potential testers.

https://github.com/FeynCalc/feynhelpers/blob/looptools/Interfaces/LoopTools.m

If you are interested, I could try to revamp it over the weekend.

**Reply by tonymenzo** [reply]
`OptionsPattern[]` didn't fix the recursive issue but I think it's a good idea to have it in there generally.

If you're up to it yeah, that looks like it would be great and prove quite useful for me. I'd be happy to put it through the wringer.

**Reply by vsht** [reply]
Now that I had a look at this again, I remember that FeynCalc and LoopTools cannot be naively loaded on the same kernel:
as both define same symbols (e.g. A0, B0 etc.) you get all kind of weird shadowing errors. You would see these errors
if you wouldn't apply Quiet to Install. BTW, if you get any errors/warnings in conjunction with FeynCalc, in 99% of cases
this signals that something gone wrong and you will get garbage as output. So it's a very bad idea to ignore those.

If you save the output of FeynCalc into a separate file and then load it on a fresh kernel (without loading FeynCalc there) with LoopTools, this should probably work.

I'm attaching a short example for using the LoopTools interface from the dev version of FeynHelpers.

https://github.com/FeynCalc/feynhelpers#development-version

See if it works for you. However, please keep in mind that LoopTools has a lot of options and I'm not an expert on that. You can
set them via the interface functions, though.

[LoopToolsExample.zip](https://github.com/FeynCalc/feyncalc/files/6216228/LoopToolsExample.zip)

**Comment by alessandrovalenti** [comment]
Hello @vsht , sorry to open again this old discussion. I need to evaluate numerically some B0,C0,D0 functions obtained through Feyncalc. I downloaded the development version of FeynHelper, however it is not working on my system, I think because I am on Windows.
Are there any way to evaluate the result numerically through looptools beside installing Linux? For instance, I tried copypasting the result in a new notebook in which I loaded only LoopTools, however because of the different output representation (B_0 in Feyncalc, B0 in LoopTools for instance) I cannot make evaluate the result. Do you have any suggestions?
Thanks, 
Alessandro

**Reply by vsht** [reply]
Hi Alessandro,

if you can install and run standalone LoopTools on your system, then the FeynHelpers interface should work as well.

LoopTools organizes the communication between the Fortran backend and the Mma frontend via a MathLink binary,
which is OS-specific.

So far I've been only testing my LoopTools interface on Linux, where the binary is expected to be located in
`/home/$USERNAME/.Mathematica/Applications/LoopTools/x86_64-Linux/bin/LoopTools`.

On Windows the binary is called `LoopTools.exe` and should be somewhere in the folder to which you installed
your LoopTools copy. I don't know the exact location, but I tried compiling LoopTools from scratch using
Cygwin on my W10 Virtual machine and upon giving `LToolsLoadLibrary` the correct path to `LoopTools.exe` everything worked
as expected.

![Bildschirmfoto vom 2021-09-01 12-02-40](https://user-images.githubusercontent.com/7283188/131654397-da2e784d-c365-4c5d-a838-d1d3d1c3a1f0.png)


So it is just the issue of finding the correct path and evaluating
```
LToolsLoadLibrary[LToolsPath ->  "C:\\PATHTOLOOPTOOLS\\LoopTools.exe"]
```
before you can use the interface.

Cheers 
Vladyslav

**Reply by alessandrovalenti** [reply]
Thank you very much @vsht : it was a path issue, I used single instead of double slashes...now I can load LoopTools.

However what I find now is a problem similar to @tonymenzo 's one: the evaluation of Fermion Spin Sum gives many errors. 
Let me explain what I am doing:


1) Fix the kinematics 
2) Write down the amplitude and use TID[ampl, ToPaVe->True]
3) Compute the conjugated amplitude
4) Fix numerical values for all the parameters to see if LoopTools works
5) LToolsEvaluate the amplitudes. Here an error appears saying that I have tensor integrals, which is of course not true after TID. However this seems not to be a problem because it ignores the vectors and evaluates only the PaVe's
6) Contract amplitude x conjugated amplitude x metric tensor. Here I have a problem because some scalar products are not evaluated
7) FermionSpinSum the result above. Here I have the recursion errors I was talking about.
I'm attacching the Mathematica notebook with these operations:
[Example.zip](https://github.com/FeynCalc/feyncalc/files/7092143/Example.zip)


I made many tries and I should add that the problem arises _only_ when I have both slashed vectors sandwhiched between the spinors _and_ vectors with a free index, ex: 
SpinorVBarD[p2, 0] . GSD[q] . SpinorUD[p1, 0]*FVD[l, \[Mu]] *denominator: NOT OK
SpinorVBarD[p2, 0]  . SpinorUD[p1, 0]*FVD[l, \[Mu]]*denominator: OK
SpinorVBarD[p2, 0] . GSD[q] . SpinorUD[p1, 0]*denominator: OK
SpinorVBarD[p2, 0] . GAD[\[Nu]] . SpinorUD[p1, 0]*FVD[l, \[Mu]]*denominator: OK

I really cannot understand what is happening. This is very important to me because I am computing amplitudes whose PaVe's functions are too complicated to handle analytically (I tried using PaXEvaluate, but Mathematica gets stuck in computing //Contract for the squared amplitudes. These amplitude are not the one in the example I am attacching, it is just a part of them), so I am thinking to evaluate the amplitude numerically (as I am doing) and then squared it. Unfortunately I cannot LToolsEvaluate directly the squared one because it contains products of PaVe's which are not recognized by both LToolsEvaluate or PaXExpand. Your help is much appreciated. 

Also, I would like to ask @tonymenzo if he solved is problem, and in case to tell me how he did, since it is certaintly valuable help to me.

Sorry for the long wall of text,
Alessandro

**Reply by vsht** [reply]
First of all let me clarify one thing: Do you really want to calculate a **1-loop amplitude squared**?

I'm asking this question because the situation in your notebook never happens in 1-loop calculations: 
the 1-loop amplitude you calculate can be multiplied only with the corresponding tree-level amplitude,
so that you never have a product of two PaVe functions. 

For example, here

https://feyncalc.github.io/FeynCalcExamplesMD/QED/OneLoop/ElAel-MuAmu

one first evaluates the 1-loop virtual correction and then multiplies it with the Born contribution.
This way one obtains the Born-virtual interference term for the muon production at 1-loop.

In a 2-loop calculation you indeed would need to consider the product of two 1-loop amplitudes (as one
of the ingredients), but for that LoopTools obviously would not be the right tool: It contains only
O(ep^0) pieces of 1-loop integrals, while for a 2-loop amplitude you need 1-loop results up to
O(ep^2) (for the case that the other 1-loop integral has a 1/ep^2 IR divergence ).

PS Of course FeynCalc has D-dimensional polarization vectors and can do the correct sums in
D-dimensions. This might not be properly documented, but to get a D-dim polarization vector
when entering amplitudes by hand, you can use e.g. ChangeDimension[PolarizationVector[p, mu], D].
If you use FeynArts, the FCFAConvert function will handle that for you if you demand to convert
the amplitude to D-dimensions.

Cheers,
Vladyslav

**Reply by tonymenzo** [reply]
Hi Alessandro, 

I vaguely remember the recursion error was related to `FermionSpinSum[]` not liking input with a floating-point coefficient attached to the spinor chain. FeynCalc isn't designed for numerical computations so the parse trees aren't written to account for floating-point coefficients. There are a number of "quick" solutions to this problem. The quickest/easiest that comes to mind would be to write a short function which parses the input to acquire the numerical coefficient and then uses the option `ExtraFactor` within `FermionSpinSum[]`. 

For example the function would convert the input
```
(1.0+1.0i)*SpinorUBar[k1, m].GS[p].GA[5].SpinorU[p1, m] -> FermionSpinSum[SpinorUBar[k1, m].GS[p].GA[5].SpinorU[p1, m], ExtraFactor->1.0+1.0i]
```
@vsht may have a better solution.

Cheers, 
Tony

**Reply by alessandrovalenti** [reply]
Thanks @vsht and @tonymenzo for your answer. 

To tony, indeed I was thinking the problem might be related to the floating-point coefficient of the spinor-chain. Thanks for the code-input, I will proceed in this direction.

To Vladyslav, I am computing a particular process involving SM+BSM particles which appears only at 1-loop (tree level contributions are pure SM).  I would like to compute (BSM)^2 because SM contribution is very small, and so will be also the interference term BSM - SM.  The integral must not be either UV or IR divergent, so I thought LoopTools should be able to manage it. However, as you say, I might wanto to look at other tools. Thanks for your anwser (and for the tip about polarization vectors in D-dimensions - indeed in the documentaiton there is only the 4-dimensional one https://feyncalc.github.io/FeynCalcBook/ref/PolarizationVector.html).

Thank you all again,
Alessandro

**Reply by vsht** [reply]
Ok, since I assume that you understand what you are doing, that's fine with me.

Still, for `LToolsEvaluate` I would like to assume that the input expressions is an amplitude linear
in `PaVe` functions. Your case of evaluating PaVe^2 that all turn out to be IR and UV finite seems to 
be too special to deserve a dedicated treatment. So let us focus on evaluating the amplitude first
and squaring thereafter.

Indeed, as Tony correctly guessed, most FeynCalc functions will freak out when given input containing floating
point numbers. So we better wrap the evaluated PaVe functions with some arbitrary head and keep them
in that state until the very end when the amplitude squared is a pure c-number made of scalar products,
evaluated integrals and other constants. To that end I added a new option Head to LToolsEvaluate. This way your
example seems to work fine

[Example2.zip](https://github.com/FeynCalc/feyncalc/files/7099799/Example2.zip)

If you want to compare numerical results to the output of Package-X, I would recommend using
the newly introduced [FCCompareNumbers](https://github.com/FeynCalc/feyncalc/blob/master/FeynCalc/DocumentationFiles/Mathematica/Shared/Tools/FCCompareNumbers.m) routine.

As to polarization sums, I'll try to update the documentation to include some d-dim examples in due time.

Cheers,
Vladyslav

**Reply by alessandrovalenti** [reply]
Thank you Vladyslav, and sorry for the delay of the answer. Indeed the functionality that you have implemented works well! This solves my problems. Thank you again, and have a good day!

**Reply by vsht** [reply]
Nice to hear that. I've updated the documentation pages on `PolarizationVector`, `PolarizationSum` and `DoPolarizationSums`
to mention `D`-dimensional polarizations. You can access them from the overview

https://github.com/FeynCalc/feyncalc/blob/master/FeynCalc/DocumentationFiles/Markdown/Extra/FeynCalc.md

of the current frugal documentation (due to reasons explained in #59, the documentation in FC 10 will be completely 
markdown based and built out of pages in DocumentationFiles/Markdown)

**Reply by alessandrovalenti** [reply]
Many thanks! If you do not mind, let me ask a last question. As I was stating some messages ago, my amplitude is both UV and IR finite. Indeed if I compute the cross section (i.e. 1-loop amplitude squared, since is a process not present at tree-level) analytically, I get a decent result: a "compact" expression written in terms of the functions C0^2 , C0xD0, D0^2, with all the particles massive. At this point it would be very nice if I could simply expand these functions analytically (via package x) or numerically (LoopTools), avoiding some of the problems encountered before. However PaXEvaluate and LToolsEvaluate do not handle C0^2, C0xD0 and D0^2. At the moment I am looking at the file Interfaces/PackageX.m in order to enable this functionality: indeed package-x alone handles the square of PaVe functions, through LoopRefine[]. But I do not have any experience in advanced Mathematica programming. So my question is: can you point me to what needs to be modified in order for this to work?
Thank you and good evening.

**Reply by vsht** [reply]
I'm afraid that the Package-X interface would require too many modifications to be able to handle squared 1-loop integrals.
Essentially, it would be probably easier to rewrite it from scratch.

In a such situation I would probably try to do all loop-related expansions on the amplitude level and then expand again when needed upon having calculated the squared amplitude.
